onnx/defs/tensor/old.cc

```diff
@@ -426,40 +426,63 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Constrain input 'X' and output 'Y' to all tensor types.")
         .SetDoc(Upsample_ver9_doc)
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
-          if (!hasNInputShapes(ctx, 1)) {
+          if (!hasNInputShapes(ctx, 2)) {
             return;
           }
           propagateElemTypeFromInputToOutput(ctx, 0, 0);
           auto& input_shape = getInputShape(ctx, 0);
           auto* output_shape = getOutputShape(ctx, 0);
-          output_shape->clear_dim();
           auto scales = ctx.getInputData(1);
+
+          if (output_shape->dim_size() > 0) {
+            if (output_shape->dim_size() != input_shape.dim_size()) {
+              fail_shape_inference(
```

Comment:
Since we support opset 7 too - I guess this fix needs to be applied to `Upsample (7)` as well. Since Upsample (7) uses attributes to store 'scales ' (scales is not an input), I would make the function handle attribute 'scales' as well and just reference that function in a ll 4 places - Upsample (7), Upsample (9), Upsample (10), Resize (10).